### PR TITLE
Role List Table component

### PR DESCRIPTION
**Pull request by jerabekjiri**
No-Issue

Adding prototype component for basic Role List Table component that is identical on new/changed `RBAC` pages.

I decided to use a functional component and use hooks (`useState`). I've seen it's used for `collection signing` (@himdel is that okay? I can revert it to class component).

Use the component in code:
```
<RoleListTable
    roles={roles}
    dropdownActions={[
        <DropdownItem
            key='edit-role'
            onClick={(role) => {
                // edit role action
            }}
        >
            {t`Edit Role`}
        </DropdownItem>,
        <DropdownItem
            key='duplicate-role'
            onClick={(role) => {
                // duplicate role action
            }}
        >
            {t`Duplicate role`}
        </DropdownItem>,
        <DropdownItem
            key='delete-role'
            onClick={(role) => {
            // delete role action
            }}
        >
          {t`Delete role`}
        </DropdownItem>,
    ]}
    params={params}
    updateParams={(p) => {
        this.updateParams(p, () => this.queryGroup());
    }}
    isUserManagementDisabled={isUserMgmtDisabled}
/>
 ```
### Expanded table 
Screenshot from 2022-03-14 15-16-02
### Dropdown actions
Screenshot from 2022-03-14 15-16-14
### Dropdown actions 2
Screenshot from 2022-03-14 15-19-08

 Would this work for your `Role Page`? @ShaiahWren

### CODE REVIEW

File: src/components/role-list-table/role-list-table.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  roles,
+  isUserManagementDisabled,
+  dropdownActions,
+  params,
+  updateParams,
+}: Props) => {
+  const [expandedRoles, setExpandedRoles] = useState<string[]>([]);
+
+  const isRoleExpanded = (role) => expandedRoles.includes(role);
+
```
Comment (ShaiahWren) [inline src/components/role-list-table/role-list-table.tsx:45]: `const [isExampleCompact, setIsExampleCompact] = React.useState(true);`

Comment (ShaiahWren) [reply src/components/role-list-table/role-list-table.tsx:45]: Shall we add this piece of state to have the option of a compact table? Tina and Taufique asked if this was possible in one of our team meetings about the refactor.

File: src/components/role-list-table/role-list-table.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      {
+        title: '',
+        type: 'none',
+        id: 'kebab',
+      },
+    ],
+  };
+
+  return (
+    <TableComposable aria-label='role-list-table' data-cy='RoleListTable'>
```
Comment (ShaiahWren) [inline src/components/role-list-table/role-list-table.tsx:91]:
```suggestion
    <TableComposable variant={isRoleCompact ? 'compact' : undefined} aria-label='role-list-table' data-cy='RoleListTable' >
```

File: src/components/role-list-table/role-list-table.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+              expand={{
+                isExpanded: isRoleExpanded(name),
+                onToggle: () => handleToggleRole(name),
+                rowIndex: i,
+              }}
+            />
+            <Td>{name}</Td>
+            <Td>{description}</Td>
+            {dropdownActions && (
+              <Td style={{ paddingRight: '0px', textAlign: 'right' }}>
```
Comment (ShaiahWren) [inline src/components/role-list-table/role-list-table.tsx:110]: This styling will be removed with the ListItemActions component on https://github.com/ansible/ansible-hub-ui/pull/1737, but no need to worry about that now :)

**Comment by jerabekjiri** [comment]
#### Added new prop `variant` (`expandable`, `radio`, `checkbox`) 
- Checkbox variant
Screenshot from 2022-03-16 22-18-35

- Radio variant
Screenshot from 2022-03-16 22-19-12

**Comment by jerabekjiri** [comment]
@ShaiahWren Looking at [AAH-1129](https://issues.redhat.com/browse/AAH-1129), I think we will need both radio and checkboxes. Radio buttons for selecting a role and checkbox for adding roles to the group.
HUB-RBAC-Namespaces-Group List View-Wizard Step 1-Selected
 
HUB-RBAC-Add Wizard-Step 1-Filter

**Comment by jerabekjiri** [comment]
- Adding `isStickyHeader` prop, which could be useful in longer list for example in `WizardModal` 
Screenshot from 2022-03-21 12-54-19

**Comment by himdel** [comment]
> I decided to use a functional component and use hooks (useState). I've seen it's used for collection signing (@himdel is that okay? I can revert it to class component).

It's OK, also discussed with David, if we're comfortable using functional components for new stuff, nothing is stopping us :)

---

Otherwise LGTM :+1:, not seeing any issues,

perhaps just one confusion .. I'm not sure the expandable role table is the same table we'd be using in the wizard?
(I'm pretty sure it is the one we'd use on the detail screen, but I think the table of groups with radios, and roles with checkboxes is a different table from the one here.)

**Comment by jerabekjiri** [comment]
### Reworked the RoleListTable
moved the logic from the component and added `ExpandableRow`, `CheckboxRow` and `RadioRow`, which can be used instead of the `variant` prop.

- Radio variant
```tsx
<RoleListTable
  isStickyHeader={true}
  params={this.state.params}
  updateParams={this.updateParams}
  isCompact={true}
>
  {roles.map((role, i) => (
    <RadioRow
      key={role.name}
      isSelected={this.state.selectedRole === role.name}
      onSelect={() =>
        this.setState({
          selectedRole: role.name,
        })
      }
    >
      <td>{role.name}</td>
      <td>{role.description}</td>
    </RadioRow>
  ))}
</RoleListTable>
```
- Checkbox variant
```tsx
<RoleListTable
  isStickyHeader={true}
  params={this.state.params}
  updateParams={this.updateParams}
  isCompact={true}
>
  {roles.map((role, i) => (
    <RadioRow
      key={role.name}
      isSelected={this.state.selectedRoles.includes(role.name)}
      onSelect={() =>
        this.setState({
          selectedRoles: [...this.state.selectedRoles, role.name],
        })
      }
    >
      <td>{role.name}</td>
      <td>{role.description}</td>
    </RadioRow>
  ))}
</RoleListTable>
```

- and expandable variant with permissions chips
```tsx
<RoleListTable
  isStickyHeader={true}
  params={this.state.params}
  updateParams={this.updateParams}
  isCompact={true}
>
  {roles.map(role => (
    <ExpandableRow
      key={role.name}
      expandableRowContent={
        <>
          {groups.map((group) => (
            <Flex
              style={{ marginTop: '16px' }}
              alignItems={{ default: 'alignItemsCenter' }}
              key={group.name}
              className={group.name}
            >
              <FlexItem style={{ minWidth: '200px' }}>
                {i18n._(group.label)}
              </FlexItem>
              <FlexItem grow={{ default: 'grow' }}>
                <PermissionChipSelector
                  availablePermissions={group.object_permissions
                    .filter(
                      (perm) =>
                        !selectedPermissions.find(
                          (selected) => selected === perm,
                        ),
                    )
                    .map((value) =>
                      twoWayMapper(value, filteredPermissions),
                    )
                    .sort()}
                  selectedPermissions={selectedPermissions
                    .filter((selected) =>
                      group.object_permissions.find(
                        (perm) => selected === perm,
                      ),
                    )
                    .map((value) =>
                      twoWayMapper(value, filteredPermissions),
                    )}
                  menuAppendTo='inline'
                  multilingual={true}
                  isViewOnly={true}
                />
              </FlexItem>
            </Flex>
          ))}
        </>
      }
    >
      <td>{role.name}</td>
      <td>{role.description}</td>
      <ListItemActions kebabItems={dropdownItems} buttons={buttons} />
    </ExpandableRow>
  }))
</RoleListTable>
```
@himdel @ShaiahWren would this work better?

File: src/components/permissions/permission-chip-selector.tsx [hunk 4]
```diff
@@ -13,7 +13,7 @@ import {
 interface IProps {
   availablePermissions: string[];
   selectedPermissions: string[];
-  setSelected: (selected: string[]) => void;
+  setSelected?: (selected: string[]) => void;
```
Comment (himdel) [inline src/components/permissions/permission-chip-selector.tsx:16]: is this bit intentional? (doesn't *seem* related?)

**Review by himdel — APPROVED** [review]
LGTM, nice separation :)

File: src/components/permissions/permission-chip-selector.tsx (hunk 4 above)
Comment (jerabekjiri) [reply src/components/permissions/permission-chip-selector.tsx:16]: yes, so I can use this component as readonly component

**Comment by jerabekjiri** [comment]
Merging, so this can unblock @ShaiahWren

**Pull request merged**